pkg/maps/fragmap/fragmap.go

```diff
@@ -0,0 +1,92 @@
+// Copyright 2020 Authors of Cilium
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package fragmap
+
+import (
+	"fmt"
+	"unsafe"
+
+	"github.com/cilium/cilium/common/types"
+	"github.com/cilium/cilium/pkg/bpf"
+	"github.com/cilium/cilium/pkg/option"
+)
+
+const (
+	// MapName is the name of the map used to retrieve L4 ports associated
+	// to the datagram to which an IPv4 belongs.
+	MapName = "cilium_ipv4_frag_datagrams"
+	// MaxEntries is the number of entries in the LRU map
+	MaxEntries = 8192
+)
+
+// FragmentKey must match 'struct ipv4_frag_id' in "bpf/lib/ipv4.h".
+// +k8s:deepcopy-gen=true
+// +k8s:deepcopy-gen:interfaces=github.com/cilium/cilium/pkg/bpf.MapKey
+type FragmentKey struct {
+	destAddr   types.IPv4 `align:"daddr"`
+	sourceAddr types.IPv4 `align:"saddr"`
+	id         uint16     `align:"id"`
+	proto      uint8      `align:"proto"`
+	pad        uint8      `align:"pad"`
+}
+
+// FragmentValue must match 'struct ipv4_frag_l4ports' in "bpf/lib/ipv4.h".
+// +k8s:deepcopy-gen=true
+// +k8s:deepcopy-gen:interfaces=github.com/cilium/cilium/pkg/bpf.MapValue
+type FragmentValue struct {
+	sourcePort uint16 `align:"sport"`
+	destPort   uint16 `align:"dport"`
+}
+
+// GetKeyPtr returns the unsafe pointer to the BPF key.
+func (k *FragmentKey) GetKeyPtr() unsafe.Pointer { return unsafe.Pointer(k) }
+
+// GetValuePtr returns the unsafe pointer to the BPF value.
+func (v *FragmentValue) GetValuePtr() unsafe.Pointer { return unsafe.Pointer(v) }
+
+// String converts the key into a human readable string format.
+func (k *FragmentKey) String() string {
+	return fmt.Sprintf("%s --> %s, %d, %d", k.sourceAddr, k.destAddr, k.proto, k.id)
+}
+
+// String converts the value into a human readable string format.
+func (v *FragmentValue) String() string {
+	return fmt.Sprintf("%d, %d", v.destPort, v.sourcePort)
+}
+
+// NewValue returns a new empty instance of the structure representing the BPF
+// map value.
+func (k FragmentKey) NewValue() bpf.MapValue { return &FragmentValue{} }
+
+// InitMap creates the signal map in the kernel.
+func InitMap() error {
+	if !option.Config.EnableIPv4 || !option.Config.EnableIPv4FragmentsTracking {
+		return nil
+	}
+
+	fragMap := bpf.NewMap(MapName,
+		bpf.MapTypeLRUHash,
+		&FragmentKey{},
+		int(unsafe.Sizeof(FragmentKey{})),
+		&FragmentValue{},
+		int(unsafe.Sizeof(FragmentValue{})),
+		MaxEntries,
+		0,
+		0,
+		bpf.ConvertKeyValue,
+	)
+	_, err := fragMap.OpenOrCreate()
```

Comment:
I think this should be:
```suggestion
	err := fragmap.Create()
```
We're not using this map in the daemon so we can close the fd.

Reply:
Ok. I kept it open mostly to be able to see it with `cilium map list`, for debug. But I guess this is a weak argument to keep it open, I can close it.

Reply:
I don't have a strong opinion either way, but if we do that, we need to do it for all other similar maps I guess, for consistency.

Reply:
Sure, I'll go for consistency with this PR, and close the fd. We can discuss opening it for the maps in general during the next datapath call, to see if there's an interest in doing a follow-up on this.